Add TensorExpression test with multiple operations

macedo22 opened the pull request:
## Proposed changes

This PR adds a unit test for `TensorExpression`s that evaluates an expression with: addition, subtraction, contraction, an inner product, and an outer product.

Specifically, it tests the evaluation of the expression: `L_{a} = R_{ab}* S^{b} + G_{a} - H_{ba}{}^{b} * T`

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

macedo22 committed:
Add TensorExpression test with multiple operations

nilsdeppe reviewed (dismissed):
LGTM, what do you think of adding the same test but for a `DataVector` as well?

macedo22 commented:
@nilsdeppe I can do that! I'll push an update when I can

macedo22 commented:
@nilsdeppe Updated the test to also test with `DataVector`s

fmahebert commented on tests/Unit/DataStructures/Tensor/Expressions/Test_MixedOperations.cpp:
```diff
@@ -0,0 +1,114 @@
+      R(ti_a, ti_b) * S(ti_B) + G(ti_a) - H(ti_b, ti_a, ti_B) * T());
+
+  for (size_t a = 0; a < 4; a++) {
+    CHECK(actual_result_tensor.get(a) == expected_result_tensor.get(a));
```
More of a question than a change request: I would have used `CHECK_ITERABLE_APPROX(actual_result_tensor, expected_result_tensor)` because my expectation would be that the TEs might not evaluate the many terms in exactly the same order as what I'd write myself, and therefore, that the result could differ by roundoff. Is this a fair concern or do you believe you can rely on getting the same order of operations as in the TEs?

macedo22 replied:
After carefully tracing through the TE implementation, in this specific case, I believe the only difference between it and `compute_expected_result` is the associativity in the summations of contractions. For example, in computing `H_0` (i.e. after `H_{ba}{}^{b}` has been contracted to `H_a`), I believe `compute_expected_result` does `H_0 = ((H(0, 0, 0) + H(1, 0, 1)) + H(2, 0, 2)) + H(3, 0, 3)`, while `TensorContract` does `H_0 = H(0, 0, 0) + (H(1, 0, 1) + (H(2, 0, 2) + H(3, 0, 3)))`. Outside of this, I believe everything else is the same, like both implementations contracting `H` before multiplying by `T` and such.

On top of the fact that I believe the contraction summation associativity is different, another important point is that the internals of the different TE types could potentially change in the future that could affect the order of operations in a larger way. So instead of changing the associativity of the summing in the test, for the purposes of maintainability and not relying on implementation, I think it's best to change this `CHECK` to `CHECK_ITERABLE_APPROX`. I think it will be even more important to make sure I use this for a future test expression that is more complex with index reordering and nested contractions. Thanks for this suggestion!

fmahebert replied:
Makes sense - thanks for looking into this!

nilsdeppe commented on tests/Unit/DataStructures/Tensor/Expressions/Test_MixedOperations.cpp:
```diff
@@ -12,18 +12,43 @@
+
+template <typename... Ts>
+void assign_unique_values_to_tensor(
+    gsl::not_null<Tensor<DataVector, Ts...>*> tensor) noexcept {
```
`const gsl::not_null` (add `const`)

nilsdeppe commented on tests/Unit/DataStructures/Tensor/Expressions/Test_MixedOperations.cpp:
```diff
@@ -12,18 +12,43 @@
+namespace {
+template <typename... Ts>
+void assign_unique_values_to_tensor(
+    gsl::not_null<Tensor<double, Ts...>*> tensor) noexcept {
```
`const gsl::not_null` (add `const`)

nilsdeppe reviewed (commented):
LGTM! Please go ahead and squash everything, including the two `const`s

macedo22 commented:
@nilsdeppe Done!

macedo22 commented:
@nilsdeppe @fmahebert Swapped out `CHECK` for `CHECK_ITERABLE_APPROX`, squashed, and rebased. The only change I made to the files in this PR is that swap, so I'm not sure if maybe the `clang-tidy` error is occuring as a result of this swap or as a result of rebasing on `develop`. Let me know if there is something I should do to resolve this!

fmahebert commented:
@macedo22 I think the clang-tidy error is from the PCH issues on develop right now. Can you rebase on develop once PR #2782 is merged (hopefully soon)?

macedo22 commented:
@fmahebert Rebased! But it looks like the issue persists

nilsvu commented:
@macedo22 It should be fixed by #2785. Sorry for the inconvenience! It's probably OK to just ignore the clang-tidy failure in this PR if you want to avoid rebasing again :)

macedo22 commented:
@nilsleiffischer Thanks for the update! I think you're right, but I rebased anyway
